File: server/features/inquiries/inquiries-core/src/main/java/de/bennyboer/kicherkrabbe/inquiries/settings/RateLimits.java
package de.bennyboer.kicherkrabbe.inquiries.settings;

import lombok.AllArgsConstructor;
import lombok.Value;

import java.time.Duration;

import static de.bennyboer.kicherkrabbe.commons.Preconditions.notNull;
import static lombok.AccessLevel.PRIVATE;

@Value
@AllArgsConstructor(access = PRIVATE)
public class RateLimits {

    RateLimit perMail;

    RateLimit perIp;

    RateLimit overall;

    public static RateLimits of(RateLimit perMail, RateLimit perIp, RateLimit overall) {
        notNull(perMail, "Per mail rate limit must be given");
        notNull(perIp, "Per IP rate limit must be given");
        notNull(overall, "Overall rate limit must be given");

        return new RateLimits(perMail, perIp, overall);
    }

    public static RateLimits init() {
        var aDay = Duration.ofHours(24);
        
        return new RateLimits(
                RateLimit.of(2, aDay),
                RateLimit.of(2, aDay),
                RateLimit.of(20, aDay)
        );
    }

    @Override
    public String toString() {
        return "RateLimits(perMail=%s, perIp=%s, overall=%s)".formatted(perMail, perIp, overall);
    }

}

File: frontend/src/app/modules/admin/modules/inquiries/models/rate-limits.ts
import { RateLimit } from './rate-limit';
import { validateProps } from '../../../../../util';

export class RateLimits {
  readonly perMail: RateLimit;
  readonly perIp: RateLimit;
  readonly overall: RateLimit;

  private constructor(props: { perMail: RateLimit; perIp: RateLimit; overall: RateLimit }) {
    validateProps(props);

    this.perMail = props.perMail;
    this.perIp = props.perIp;
    this.overall = props.overall;
  }

  static of(props: { perMail: RateLimit; perIp: RateLimit; overall: RateLimit }): RateLimits {
    return new RateLimits(props);
  }
}
